### ISSUE TITLE
Allow adapter-specific customisations 

### ISSUE THREAD

**Issue by iMacTia** [lostisland/faraday#648]
Faraday was born as an abstraction layer between developers and their networking adapters. For this reason, when a feature is supported by Faraday, this should be also supported by ALL its adapters.
This allows you to swap your adapter freely without changing your code.

For this reason, the Faraday team is reluctant to the idea of accepting new features supported by only one or two adapters. However, I can see an increase in the number of Issues/PRs that ask for a way to access and customise adapters to support ad-hoc features.

`ActiveRecord` works in a very similar way, providing an abstraction layer with your DB adapter, that you're free to swap anytime, but with some limitations.
AR however gives you the possibility to run arbitrary SQL to query or migrate your DB, even though this could cause you to lose the possibility to use another adapter in the future (unless you change this custom code, of course).

Wanting to do something similar to AR on Faraday, we could allow to interact with the adapter to setup some custom features.
One of the most popular ways to do this, is using the block syntax:

```ruby
conn = Faraday.new(...) do |f|
  f.adapter :net_http_persistent
end

conn.configure_adapter do |adapter|
  adapter.idle_timeout = 100
end
```

This of course requires Faraday to know where the adapter is in the middleware stack, which is currently not possible, but a nice feature I would like to include in v 1.0 making the call to `#adapter` method mandatory while building the stack.

I'm happy to hear the thoughts of the community, and especially the ones from @mislav, or to hear any alternative solution on the above :)

**Comment by iMacTia** [comment]
Example of Issues/PRs that could be addressed by this feature:

#645 #585 #572 #545 #502

**Comment by mltsy** [comment]
I agree.  Restricting the library to be a lowest common denominator of all the supported adapters is an unnecessary cost.  You get the best of both worlds by consolidating any changes or incompatibilities into a configure block for a specific adapter as proposed :+1:

Faraday is actually doing 2 things:
1) Providing an easy, familiar, streamlined interface for creating and reusing HTTP connections, regardless of the underlying adapter (the reason most people use it, I imagine)
2) Providing an abstract interface that contains only the features that all supported adapters have in common, so that switching adapters is **minimally intrusive**.  Just having the configuration in a single place, and not having to change the syntax of `get` and `post` (...) statements are the biggest advantages gained here (both of which would be retained even if adapter-specific config options were allowed).  Surely any benefit is not dampened by adding the ability to explicitly break that contract if one chooses to (in a way that makes it obvious that certain settings are not compatible with other adapters).  It is still **minimally intrusive**, meaning there is no less intrusive way to change out your adapter if your use case *requires* adapter-specific configurations.

It doesn't seem like a disadvantage to say "if you have chosen to use adapter-specific configurations, you are forgoing the benefit of being able to change out your adapter with one that is incompatible with your adapter-specific use cases." In fact... that is true in *every* such case, regardless of how Faraday is constructed :)  It's just that by refusing to allow such options in Faraday, we also deprive people the other benefits of Faraday if they have such a use case, and require them to actually go find another adapter to use directly that does support their use case.

For a real-world example: In my case where I need to utilize the `verify_callback` option, I have only 2 options:
1) Use an updated version of Faraday with an adapter-specific config option to verify the SSL certificate with a custom callback, and forgo the ability to swap out the underlying adapter with one that doesn't support `verify_callback`.
2) Use `Net::HTTP` or `Excon` directly, and forgo *all* ability to easily switch out the adapter even with one that does support `verify_callback`, as well as forgoing the nice, familiar interface Faraday provides, which is already used elsewhere in my app. (This is what I have had to do, and is clearly a less ideal solution)

So... what benefit does any developer gain by not allowing these options?

**Comment by mltsy** [comment]
Would it be easier to just pass-through a certain set of attributes directly to the adapter within the initializer?  Like:

```ruby
conn = Faraday.new('<host>',
  adapter_specific_config: {
    verify_callback: method(:verify_certificate),
    idle_timeout: 100
  }
)
end
```

(Either than or in the request block) Then we don't have to know the order of the middleware stack, and the adapter-specific options are still semantically separated from the rest.

And then, in **faraday/adapter/net_http.rb** (for instance):
```ruby
def call(env)
  super
  with_net_http_connection(env) do |http|
    custom_config(http, env[:adapter_specific_config])
    ...
  end
end

def custom_config(http, options)
  options.each do |name, value|
    http.send("#{name}=", value) # Just as an easy example - or we could set them individually, if we want to restrict to known options
  end
end
```

**Comment by mltsy** [comment]
Or perhaps even better, couldn't we just yield the adapter to a block passed in the initializer for configuration?

```ruby
conn = Faraday.new(...) do |f|
  f.adapter_config do |adapter|
    # Should be easy enough to yield the adapter to this block right?
    adapter.idle_timeout = 100
  end
end

**Comment by iMacTia** [comment]
I personally don't like the meta programming solution based on the `adapter_specific_config` as it's limited to the fact that you can only assign direct properties. Having the adapter yielded into a block allows you to do anything with it, which is much more powerful.

Your last proposed solution is actually equivalent to mine as the parameter passed to the initialisation block should be the connection itself. So you can call `configure_adapter` (or `adapter_config` as you called it) inside the initialisation block, outside, or even on both places.

**Comment by mltsy** [comment]
Ooh, I see - I was thinking about that block wrong.  What I intended was to eliminate the need to know about the order of the middleware stack.  I think this is closer to what I was thinking - passing an optional second argument to the block directly from the initializer (if the block accepts 2 arguments):

```ruby
conn = Faraday.new(...) do |f, adapter|
  adapter.idle_timeout = 100
end
```

This way the adapter can just be yielded from the initializer without having to find it in the stack right? (assuming all adapter classes know which middleware is the actual "adapter" during initialization)

As for the other solution - the meta-programming is only one way to address the multiple options that could be passed in (which happens to work well for many of the options of Net:HTTP).  If we wanted to avoid that though, the custom options could be managed individually, like all the other options (the same way most PRs have been addressing them).  It doesn't have to use meta-programming.  The key here is just that those options are separated semantically into an adapter-specific section which makes it clear that they are not fully compatible options, but they would basically act just like any other option otherwise.

I agree that yielding the adapter at some point is a more robust solution though :)

**Comment by iMacTia** [comment]
@mltsy this is unfortunately not possible as you actually set the initialiser inside the initialisation block.
This would be really confusing:

```ruby
conn = Faraday.new(...) do |f, adapter|
  adapter.idle_timeout = 100 # Where did I specify which adapter I want to use?
  f.adapter = :em_http # what happens to changes done to `adapter` now that I'm setting another one?
end
```

I can see you trying to avoid knowing which middleware IS the adapter, but that's unfortunately difficult (If not impossible) to accomplish. My initial thought was to have something like:

```ruby
conn = Faraday.new(...) do |f|
  f.adapter :net_http_persistent do |a|
    a.idle_timeout = 100
  end
end
```

which I personally like, but the issue is that `adapter` is just an helper method built on top of `use` and will then pass the block to the initialiser of the adapter.
This will probably work for most of the adapters but could actually brake for others, hence my preference to go with the external block configuration.
The other advantage of the external block configuration is that you can call it whenever and wherever you want, even multiple times.

An additional thought on making the call to the `adapter` method mandatory: this is actually not strictly needed from the beginning. For example, we could rise exception ONLY if you call the `configure_adapter` method without calling `adapter`.

**Comment by mltsy** [comment]
Oooh, yes, of course.  I'm catching up... slowly... ;)  What I was trying to imitate is the way ssl options are passed to the initializer, and handled in the adapter class, and the reason it doesn't work is that the ssl options are just saved to a variable (`@ssl`) and given to the adapter when it is initialized, but the block is executed immediately.  What if we did the same thing with an adapter configuration block, and saved it as a Proc that gets passed to the adapter class on initialization?

```ruby
conn = Faraday.new(...) do |f|
  f.configure_adapter do |adapter| # gets saved as a Proc in a Connection instance variable alongside @ssl, @options, etc.
    adapter.idle_timeout = 100
  end
  f.adapter = :em_http
end
```

It does remove the ability to reconfigure the adapter after the connection is created I suppose.  But I think that's true for your ssl options, etc. anyway right?  So it seems consistent, and probably like a good idea (to create a new connection if you want to reconfigure your adapter).

**Comment by iMacTia** [comment]
It may looks consistent with ssl_options (even though I'm quite sure you can set them again, in that case the connection is closed and reopened), but you'll apply the same limitation on all other options that could be changed on-the-fly even between each request.
Having the possibility to call the configuration block multiple times puts more responsibility on developers, but I personally support it as it also grants maximum flexibility.

**Comment by mltsy** [comment]
Alright, requiring a call to adapter is looking like it might be the more straight-forward option :dizzy_face:   But my basic point was that with SSL options, we are configuring the adapter without knowing where it is in the middleware stack, so what is stopping us from doing the same thing with a Proc that configures the adapter (particularly if ssl can also be configured after creating a connection)?  [Make it available to the adapter class](https://github.com/lostisland/faraday/blob/a712938071bb0997410b5de4a4f2517d30fc4d3d/lib/faraday/rack_builder.rb#L193) and [the adapter class uses it as necessary](https://github.com/lostisland/faraday/blob/a712938071bb0997410b5de4a4f2517d30fc4d3d/lib/faraday/adapter/net_http.rb#L33). Right?

But I suppose you're right that there may be options you want to change on the fly between requests...  And requiring a call to adapter in order to use adapter-specific configurations doesn't seem like a problem either, now that you've pointed out that it wouldn't be necessary unless calling `configure_adapter` :)

**Comment by iMacTia** [comment]
I spent some time yesterday having a closer look at the rack app structure.
Unfortunately the separation of the adapter from the other middlewares appears more complicated than I expected (trying to preserve the backward-compatibility).
At this point, I'll try to implement the initial solution I came up with, soon replaced with the more powerful `configure_adapter` that, at this point, looks impossible to implement.

Solution will be something like:

```ruby
conn = Faraday.new(...) do |f|
  f.adapter :net_http_persistent do |a|
    a.idle_timeout = 100
  end
end
```

As @mltsy suggested, the block will be saved and used later when the adapter is built.
Main points from this implementation:

1. You can only specify adapter-specific configuration when setting up the connection
2. This will be implemented on a per-adapter basis, meaning that some adapter could not support this feature (but I'll try to make it available on all of them)

**Comment by iMacTia** [comment]
#651 was merged and will be released today with v0.11.0 🎉

**Comment by mltsy** [comment]
WOOHOO! :champagne: Thanks @iMacTia
